thorclient/wsclient/client.go

```diff
@@ -0,0 +1,172 @@
+}
+
+// SubscribeTxPool subscribes to pending transaction pool updates based on the provided query.
+// It returns a Subscription that streams pending transaction messages or an error if the connection fails.
+func (c *Client) SubscribeTxPool(query string) (*common.Subscription[*subscriptions.PendingTxIDMessage], error) {
+	conn, err := c.connect("/subscriptions/txpool", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.PendingTxIDMessage](conn), nil
+}
+
+// SubscribeBeats2 subscribes to Beat2 messages based on the provided query.
+// It returns a Subscription that streams Beat2 messages or an error if the connection fails.
+func (c *Client) SubscribeBeats2(query string) (*common.Subscription[*subscriptions.Beat2Message], error) {
+	conn, err := c.connect("/subscriptions/beat2", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.Beat2Message](conn), nil
+}
+
+// subscribe starts a new subscription over the given WebSocket connection.
+// It returns a read-only channel that streams events of type T.
+func subscribe[T any](conn *websocket.Conn) *common.Subscription[*T] {
+	// Create a new channel for events
+	eventChan := make(chan common.EventWrapper[*T], 1_000)
+	var closed bool
+
+	// Start a goroutine to handle receiving messages from the WebSocket connection.
+	go func() {
+		defer close(eventChan)
+		defer conn.Close()
+
+		for {
+			conn.SetReadDeadline(time.Now().Add(readTimeout))
+			var data T
+			// Read a JSON message from the WebSocket and unmarshal it into the data.
+			err := conn.ReadJSON(&data)
+			if err != nil {
+				if !closed {
+					// Send an EventWrapper with the error to the channel.
+					eventChan <- common.EventWrapper[*T]{Error: fmt.Errorf("%w: %w", common.ErrUnexpectedMsg, err)}
+				}
+				return
+			}
+
+			eventChan <- common.EventWrapper[*T]{Data: &data}
+		}
+	}()
+
+	return &common.Subscription[*T]{
+		EventChan: eventChan,
+		Unsubscribe: func() {
+			closed = true
+			conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+			conn.Close()
+		},
+	}
+}
+
+// connect establishes a WebSocket connection to the specified endpoint and query.
+// It returns the connection or an error if the connection fails.
+func (c *Client) connect(endpoint, rawQuery string) (*websocket.Conn, error) {
+	u := url.URL{
+		Scheme:   c.scheme,
+		Host:     c.host,
+		Path:     endpoint,
+		RawQuery: rawQuery,
+	}
+
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	conn.SetPingHandler(func(payload string) error {
+		// Make a best effort to send the pong message.
+		_ = conn.WriteControl(websocket.PongMessage, []byte(payload), time.Now().Add(time.Second))
+		conn.SetReadDeadline(time.Now().Add(readTimeout))
+		return nil
+	})
+	// TODO append to the connection pool
```

Comment:
Not sure this comment is possible to achieve.

Reply:
We should have a limited connection pool in the future. 
Also, we should re-use the websocket connection instead creating one per request, although that might be tricky, impossible given we're connecting sockets to specific endpoints.